internal/dispatch/graph/computecheck_test.go

```diff
@@ -369,6 +371,373 @@ func TestComputeCheckWithCaveats(t *testing.T) {
+							},
+						},
+					},
+				},
+			},
+			[]check{
+				{
+					"group:ui_apps#allowed@application:frontend_app",
+					map[string]any{
+						"provided": map[string]any{"type": "frontend", "region": "eu", "team": "shop"},
+					},
+					v1.ResourceCheckResult_MEMBER,
+					nil,
+					"",
+				},
+				{
+					"group:ui_apps#allowed@application:frontend_app",
+					map[string]any{
+						"provided": map[string]any{"type": "frontend", "region": "us"},
+					},
+					v1.ResourceCheckResult_NOT_MEMBER,
+					nil,
+					"",
+				},
+				{
+					"group:backend_apps#allowed@application:backend_app",
+					map[string]any{
+						"provided": map[string]any{
+							"type": "backend", "region": "us", "team": "shop",
+							"additional_attrs": map[string]any{
+								"tag1": 100,
+								"tag2": false,
+							},
+						},
+					},
+					v1.ResourceCheckResult_MEMBER,
+					nil,
+					"",
+				},
+				{
+					"group:backend_apps#allowed@application:backend_app",
+					map[string]any{
+						"provided": map[string]any{
+							"type": "backend", "region": "us", "team": "shop",
+							"additional_attrs": map[string]any{
+								"tag1": 200,
+								"tag2": false,
+							},
+						},
+					},
+					v1.ResourceCheckResult_NOT_MEMBER,
+					nil,
+					"",
+				},
+			},
+		},
+		{
+			"authorize if resource was created before subject",
+			`definition root {
+						relation actors: actor
+					}
+                    definition resource {
+						relation creation_policy: root#actors
+						permission tag = creation_policy
+					}
+					
+					definition actor {}`,
+			map[string]caveatDefinition{
+				"created_before": {
+					"timestamp(actor_created_at) > timestamp(created_at)",
+					map[string]caveats.VariableType{
+						"created_at":       caveats.StringType,
+						"actor_created_at": caveats.StringType,
+					},
+				},
+			},
+			[]caveatedUpdate{
+				{
+					core.RelationTupleUpdate_CREATE,
+					"resource:foo#creation_policy@root:root#actors",
+					"created_before",
+					map[string]any{
+						"created_at": "2022-01-01T10:00:00.021Z",
+					},
+				},
+				{
+					core.RelationTupleUpdate_CREATE,
+					"root:root#actors@actor:johndoe",
+					"",
+					nil,
+				},
+			},
+			[]check{
+				{
+					"resource:foo#tag@actor:johndoe",
+					map[string]any{
+						"actor_created_at": "2022-01-01T11:00:00.021Z",
+					},
+					v1.ResourceCheckResult_MEMBER,
+					nil,
+					"",
+				},
+				{
+					"resource:foo#tag@actor:johndoe",
+					map[string]any{
+						"actor_created_at": "2022-01-01T09:00:00.021Z",
+					},
+					v1.ResourceCheckResult_NOT_MEMBER,
+					nil,
+					"",
+				},
+			},
+		},
+		{
+			"time-bound permission",
+			`definition resource {
+						relation reader: user
+						permission view = reader
+					}
+					
+					definition user {}`,
+			map[string]caveatDefinition{
+				"expired": {
+					"now() < timestamp(expiration)",
+					map[string]caveats.VariableType{
+						"expiration": caveats.StringType,
```

Comment:
we have a TimestampType for dates and times: https://github.com/authzed/spicedb/blob/main/pkg/caveats/types.go#L29

Maybe add another test using it too?

Reply:
I added a test [here](https://github.com/authzed/spicedb/pull/878/files#diff-e576918c5584100ec96a2edd9fcf0b3048441257187a53378bd0ed720b910c1eR50-R67). I can change this test to use it, I just initially didn't know how to instantiate it but eventually figured it out

Reply:
We will address this in a follow up